pyro/infer/mcmc/api.py

```diff
@@ -337,18 +339,36 @@ def __init__(self, kernel, num_samples, warmup_steps=None, initial_params=None,
 
     def run(self, *args, **kwargs):
         num_samples = [0] * self.num_chains
-        z_acc = defaultdict(lambda: [[] for _ in range(self.num_chains)])
+        z_flat_acc = [[] for _ in range(self.num_chains)]
         with pyro.validation_enabled(not self.disable_validation):
             for x, chain_id in self.sampler.run(*args, **kwargs):
-                if num_samples[chain_id] == self.num_samples:
+                if num_samples[chain_id] == 0:
```

Comment:
Thanks for detecting and fixing this issue! My preliminary thoughts on this are as follows:

 - Based on your careful profiling results, I think our existing examples are all small and `.clone` does not incur much of an overhead on these examples. My guess is (correct me if I am mistaken) that it will also not matter for examples with larger sample sites because the rate determining step there would be the worker processes and not the main process that is accumulating the samples. In that scenario, I'm not sure if its worth optimizing for 1 .clone over 10 .clone on smaller tensors. If it is not, I think we might as well keep our code simpler and slightly easier to debug by not doing this flattening / unflattening. Also we can just do all the `.clone`s in the `_Multisampler` and avoid it for the the `chain=1` case. That way, even if there is some minor regression it is confined to the multi chain case.
 - If it is the case that there is a significant difference, is it possible to just make changes to do this flattening / unflattening only within the workers / _MultiSampler so that we don't need to add complexity / any additional overhead to the single chain case?

Reply:
About the first point, I thought that you would worry about the overhead issue because you did mention that there are applications which requires large number of parameters or large tensors, so I tried to think how to optimize it to avoid regression. If you don't worry about it, I will make the change to use only `.clone()` operator. My little background told me that when the sampler (generator) generates a sample, it have to wait for the consumer triggers some `next(...)` signal, then it'll execute the code after `yield ...` statement. In our case, the code after `yield` statement is to activate the subprocess which generates that sample (through `event.set()`). So if the consumer overhead is large (in case of it has to perform many `clone` operators), that subprocess has to wait a while to run again. That might not be the end of the story, subprocesses put samples to a queue. So a sample from a subprocess has to wait for current samples in the queue consumed before it is yielded to consumer.

About the second point, it would be more complicated (and harder to debug) if we maintain two separate codes for MultiSampler and SingleSampler. Consider the single chain case, I don't think that there is overhead involved here. Could you suggest an example so I can test? I did a benchmark to compare two ways
+ populate samples to a defaultdict, for each site, apply a stack operator over num_samples
+ apply num_samples flatten operator + 1 stack operator over num_samples  + split the result into num_sites
and see that they are quite comparable

```
from collections import defaultdict
import torch
params = [{'x_{}'.format(i): torch.tensor(i) for i in range(100)} for j in range(10000)]

%%time
z = defaultdict(list)
for j in range(10000):
    for i in range(100):
        z['x_{}'.format(i)].append(params[j]['x_{}'.format(i)])
stacked_params = {}
for i in range(100):
    stacked_params['x_{}'.format(i)] = torch.stack(z['x_{}'.format(i)])

%%time
flatten_params = [torch.cat([s.reshape(-1) for s in p.values()]) for p in params]
stacked_flatten_params = torch.stack(flatten_params)
stacked_params = {}
for i in range(100):
    stacked_params['x_{}'.format(i)] = stacked_flatten_params[:, i]
```
which in my system shows that the first approach took 2s and the second approach took 2.5s.

That is with num_sites=100. If I change 100 -> 10, I observe that they are 250ms vs 320ms. So I think that these numbers are quite small for us to worry about.

Reply:
>  So if the consumer overhead is large (in case of it has to perform many clone operators), that subprocess has to wait a while to run again. That might not be the end of the story, subprocesses put samples to a queue. So a sample from a subprocess has to wait for current samples in the queue consumed before it is yielded to consumer.

While this is true currently, I believe it is only an artificial constraint. In the workers, we can always put the samples into the queue and also append the sample to a list and only use `mp.Event` to periodically clear that list (which we discussed at one point). That way, the workers can generate samples more or less asynchronously without waiting for the main process to read off the same sample. I'm not sure if its worth making this optimization just yet, but we can do so if reading from the main process and cloning is expensive for a model.

> If you don't worry about it, I will make the change to use only .clone() operator.

I only prefer the simpler solution because doing otherwise seems like a case of premature optimization at this point. It is good to know that we can do this if this really turned out to be a significant problem though. That said, this is purely an internal change and does not affect the user interface, so I'll defer to your judgment on this.

Reply:
I'm also warming up to this idea of using packed/unpacked tensors. See my comment above, though you don't need to worry about building any generic utilities in this PR; I'm happy to merge this as is, and refactor this later if needed.

Reply:
I think that we can test our assumption with a simple kernel. Will do it soon to see the effect of these changes.

Reply:
> I think that we can test our assumption with a simple kernel.

Please don't spend too much time profiling. I don't think there are any perf repercussions to this change, and my concerns around code complexity can be handled by building generic utilities later.

Reply:
I just test on a simple PriorKernel
```
class PriorKernel(MCMCKernel):
    @property
    def initial_params(self):
        return self._initial_params

    @initial_params.setter
    def initial_params(self, params):
        self._initial_params = params

    def sample(self, params):
        return params

kernel = PriorKernel()
initial_params = {'x_{}'.format(i): torch.tensor(i).repeat(4) for i in range(100)}
mcmc = MCMC(kernel, 1000, 0, initial_params=initial_params, num_chains=4, transforms={})
mcmc.run()
```
and see that the dev version took 1m24s with the default sharing strategy and 4s with `torch.multiprocessing.set_sharing_strategy('file_system')` (which is default in MacOS) while this PR took 1s. Beyond the memory error, it is also easy to trigger `RuntimeError: received 0 items of ancdata.` with default sharing strategy (for example, baseball example). So we can see other benefits of using this pack/unpack job:
+ users don't have to worry about the error `RuntimeError: received 0 items of ancdata.` with default sharing strategy
+ it is a bit faster when model has many sites :)

Reply:
I think with the PriorKernel, the workers will be super fast, so it may not be surprising that with large number of sites, we are bottlenecked in the main process but the perf benefit might not be noticeable for HMC/NUTS. :)

> torch.multiprocessing.set_sharing_strategy('file_system')

Should we make this the default?

Reply:
> Should we make this the default?

It just happens when we don't use `clone` & `del`. Because this PR is merged, I think we don't have to worry about it.

I forgot the test for using `.clone()` for each site. It took 18s (comparing to 4s without clone and 1s when using pack/unpack) to get 1000 samples. Delay the sampling speed (by using `time.sleep(0.01)`) does not affect pack/unpack pattern (took 11s to get 1000 samples) but it helps for the case we use `clone` for each site (took 19s to get 1000 samples). So somehow the overhead of the main process does not affect much the sampling speed.